Key identity

VldMrgnn opened the pull request:
Created a function to sort the keys of option object to ensure the identity of the "key" property of actions payload, on same endpoint with same parameters.  
We have to be able to recreate the key when necessary based on the properties of the action.
The proposed function orders recursively the keys of an object. In case of primitive values   it just settles to an object to get stringified later on.
e.g.
```js
koSort(1) = {opts:1}
```
In case of an array it doesn't sort the elements of the array but it sorts the keys within the elements of the array  if they are objects.
The sort function will not alter the way the options are set. It prepares the object to be stringified along with the name of the action but will not interfere with the option object as it was passed.

neurosnap commented on src/pipe.ts:
```diff
@@ -138,12 +142,31 @@ export function createPipe<Ctx extends PipeCtx = PipeCtx<any>>({
+          .sort()
+          .reduce((res: any, key: any) => {
+            res[`${key}`] = opts[key];
+            if (opts[key] && typeof opts[key] === 'object') {
```
Is there a reason why you didn't use `isObject` here?

VldMrgnn replied:
No. I just observed the custom "isObject"  function later.

neurosnap commented on src/pipe.ts:
```diff
@@ -138,12 +142,31 @@ export function createPipe<Ctx extends PipeCtx = PipeCtx<any>>({
+
     const actionFn = (options?: any) => {
-      const key = encodeBase64(JSON.stringify({ name: createName, options }));
+      const koSort = (opts?: any): object => {
```
I'd like to see this function moved to a separate file: `create-key.ts` which also includes the key creation functionality.

VldMrgnn replied:
I felt that this was the better approach as well but I didn't dare to create new files just for this small function.

neurosnap commented on src/pipe.test.ts:
```diff
@@ -351,3 +353,334 @@ test('middleware order of execution', async (t) => {
   t.assert(acc === 'abcdefg');
 });
+
+const getKeyOf = (action: ActionWithPayload<any> | Action): string =>
```
Please move these tests to `create-key.test.ts`

neurosnap reviewed (changes requested):
Thanks for submitting this PR!  I left some comments, happy to answer any follow up questions you may have!

Another request: please remove the `package-lock.json` file.

VldMrgnn commented on package.json:
```diff
@@ -38,7 +38,7 @@
   "devDependencies": {
     "@types/node": "^17.0.23",
-    "ava": "^4.1.0",
+    "ava": "^4.2.0",
```
No reason. Reverted to original 4.1.0

VldMrgnn commented on src/fetch.test.ts:
```diff
@@ -8,7 +8,7 @@ import { requestMonitor } from './middleware';
 const mockUser = { id: '1', email: 'test@saga-query.com' };
 
-const delay = (n: number = 50) =>
+const delay = (n: number = 200) =>
```
Yes. This needs some attention. As I run each tests separately all passed. But when I run ```yarn test``` then randomly at about one of each 10 runs it did not passed  "fetch - should be able to fetch a resource and save automatically" or "fetch - POST".
After some  tests and debugging I finally found out that increasing the delay solves the problem.
Please see below:
```js

const delay = (n: number = 50) =>
  new Promise((resolve) => {
    setTimeout(resolve, n);
  });

test('fetch - should be able to fetch a resource and save automatically', async (t) => {
  t.plan(3);

  nock(baseUrl).get('/users').reply(200, mockUser);

  const api = createApi();
  api.use(requestMonitor());
  api.use(api.routes());
  api.use(fetcher({ baseUrl }));

  const fetchUsers = api.get('/users', function* (ctx, next) {
    ctx.cache = true;
    yield next();

    t.deepEqual(ctx.request, {
      url: `${baseUrl}/users`,
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
      },
    });

    t.deepEqual(ctx.json, { ok: true, data: mockUser });
  });

  const store = setupStore(api.saga());
  const action = fetchUsers();
  store.dispatch(action);

  await delay();

  const state = store.getState();
  t.log(state);
  t.deepEqual(state['@@saga-query/data'][action.payload.key], mockUser);
});

```
This code logs nonrmally:
```json
 ℹ {
        '@@saga-query/data': {
          'eyJuYW1lIjoiL3VzZXJzIFtHRVRdIn0=': {
            email: 'test@saga-query.com',
            id: '1',
          },
        },
        '@@saga-query/loaders': {
          '/users [GET]': {
            lastRun: 1652606932688,
            lastSuccess: 1652606932741,
            message: '',
            meta: {},
            status: 'success',
          },
        },
      }
```
but on 5-th run in this case (on yarn test, not individually) fasils loging
```json

 ℹ {
        '@@saga-query/data': {},
        '@@saga-query/loaders': {
          '/users [GET]': {
            lastRun: 1652606948080,
            lastSuccess: 0,
            message: '',
            meta: {},
            status: 'loading',
          },
        },
      }
```
I found out that increasing the delay solves the issue. 
I didn't really knew what to do in such situation (if I should comment here or not)

VldMrgnn replied:
I provide a clearer log output (which errored on 17-th run, not before):

```js
test('fetch - should be able to fetch a resource and save automatically', async (t) => {
  t.plan(3);

  nock(baseUrl).get('/users').reply(200, mockUser);

  const api = createApi();
  api.use(requestMonitor());
  api.use(api.routes());
  api.use(fetcher({ baseUrl }));

  const fetchUsers = api.get('/users', function* (ctx, next) {
    ctx.cache = true;
    yield next();

    t.deepEqual(ctx.request, {
      url: `${baseUrl}/users`,
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
      },
    });

    t.deepEqual(ctx.json, { ok: true, data: mockUser });
  });

  const store = setupStore(api.saga());
  const action = fetchUsers();
  t.log('action', action)
  store.dispatch(action);

  await delay();

  const state = store.getState();
  t.log(state);
  t.deepEqual(state['@@saga-query/data'][action.payload.key], mockUser);
});

```
resulting:

```perl
 ✖ fetch › fetch - should be able to fetch a resource and save automatically 
    ℹ action {
        payload: {
          key: 'eyJuYW1lIjoiL3VzZXJzIFtHRVRdIn0=',
          name: '/users [GET]',
          options: undefined,
        },
        type: '@@saga-query/users [GET]',
      }
    ℹ {
        '@@saga-query/data': {},
        '@@saga-query/loaders': {
          '/users [GET]': {
            lastRun: 1652607552471,
            lastSuccess: 0,
            message: '',
            meta: {},
            status: 'loading',
          },
        },
      }
```

VldMrgnn commented:
Hello. I just updated the library to latest version 4.0.0.
As there are changes of how the code in organized, I think that we better just drop this PR and I should create another, updated.  So, please advise on that.
I have to ask if your intention is to have the same key for the falsy primitive parameters. Because if so, I will update the function and the tests accordingly. 
For example ```dispatch(someApi(0))```  would yield the same key as ```dispatch(someApi())```  according to the latest approach. 
If not, my guess would be to  return an encoded part of key at least for  ```0``` and ```false```, but in such case why not for any falsy value.

Regarding the [issue of the "delay"](https://github.com/neurosnap/saga-query/pull/13#discussion_r873091286) from above, the testing on 4.0.0 behaves the same. As I increase the delay to 200 there are no more errors on testing on 'yarn test'. 

Thank you, Vlad

neurosnap commented:
Yea sorry about the potential conflicts with my latest version. If you want to create a new PR that is fine with me, totally up to you.

VldMrgnn commented:
> Yea sorry about the potential conflicts with my latest version. If you want to create a new PR that is fine with me, totally up to you.

Nothing to be sorry about! I am glad to do it and I will come up with a new PR.
